soc/renesas/ra/ra4m1/soc.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2024 Ian Morris
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/kernel.h>
+
+#if MHZ(24) == CONFIG_SOC_OPTION_SETTING_HOCO_FREQ
+#define OFS1_HOCO_FREQ		0
+#elif MHZ(32) == CONFIG_SOC_OPTION_SETTING_HOCO_FREQ
+#define OFS1_HOCO_FREQ		2
+#elif MHZ(48) == CONFIG_SOC_OPTION_SETTING_HOCO_FREQ
+#define OFS1_HOCO_FREQ		4
+#elif MHZ(64) == CONFIG_SOC_OPTION_SETTING_HOCO_FREQ
+#define OFS1_HOCO_FREQ		5
+#else
+#error "Unsupported HOCO frequency"
+#endif
+
+#if 3840 == CONFIG_SOC_OPTION_SETTING_VDSEL
+#define OFS1_VDSEL			0
+#elif 2820 == CONFIG_SOC_OPTION_SETTING_VDSEL
+#define OFS1_VDSEL			1
+#elif 2510 == CONFIG_SOC_OPTION_SETTING_VDSEL
+#define OFS1_VDSEL			2
+#elif 1900 == CONFIG_SOC_OPTION_SETTING_VDSEL
+#define OFS1_VDSEL			3
+#elif 1700 == CONFIG_SOC_OPTION_SETTING_VDSEL
+#define OFS1_VDSEL			4
+#else
+#error "Unsupported voltage detection level"
+#endif
+
+struct ofs0_reg {
+	uint32_t rsvd1: 1;
+	uint32_t IWDTSTRT: 1;
+	uint32_t IWDTTOPS: 2;
+	uint32_t IWDTCKS: 4;
+	uint32_t IWDTRPES: 2;
+	uint32_t IWDTRPSS: 2;
+	uint32_t IWDTRSTIRQS: 1;
+	uint32_t rsvd2: 1;
+	uint32_t IWDTSTPCTL: 1;
+	uint32_t rsvd3: 2;
+	uint32_t WDTSTRT: 1;
+	uint32_t WDTTOPS: 2;
+	uint32_t WDTCKS: 4;
+	uint32_t WDTRPES: 2;
+	uint32_t WDTRPSS: 2;
+	uint32_t WDTRSTIRQS: 1;
+	uint32_t rsvd4: 1;
+	uint32_t WDTSTPCTL: 1;
+	uint32_t rsvd5: 1;
+};
+
+struct ofs1_reg {
+	uint32_t rsvd1: 2;
+	uint32_t LVDAS: 1;
+	uint32_t VDSEL1: 3;
+	uint32_t rsvd2: 2;
+	uint32_t HOCOEN: 1;
+	uint32_t rsvd3: 3;
+	uint32_t HOCOFRQ1: 3;
+	uint32_t rsvd4: 17;
+};
+
+struct mpu_regs {
+	uint32_t SECMPUPCSO;
+	uint32_t SECMPUPCEO;
+	uint32_t SECMPUPCS1;
+	uint32_t SECMPUPCE1;
+	uint32_t SECMPUS0;
+	uint32_t SECMPUE0;
+	uint32_t SECMPUS1;
+	uint32_t SECMPUE1;
+	uint32_t SECMPUS2;
+	uint32_t SECMPUE2;
+	uint32_t SECMPUS3;
+	uint32_t SECMPUE3;
+	uint32_t SECMPUAC;
+};
+
+struct opt_set_mem {
+	struct ofs0_reg ofs0;
+	struct ofs1_reg ofs1;
+	struct mpu_regs mpu;
+};
+
+#ifdef CONFIG_SOC_OPTION_SETTING_MEMORY
+const struct opt_set_mem ops __attribute__((section(".opt_set_mem"))) = {
+	.ofs0 = {
+		/*
```

Comment:
I couldn't estimate what the design intention from the "Set all fields to 1” description.
Since WDTSTPCTL is 1, the watchdog will not start, so it would be better to state that.